src/pooler.erl

```diff
@@ -662,21 +664,35 @@ do_return_member(Pid, ok, #pool{name = PoolName,
             Fmt = "pool '~s': ignored return of free member ~p",
             error_logger:warning_msg(Fmt, [PoolName, Pid]),
             Pool;
+        %% Connection Pid is attached to a consumer process.
         {ok, {MRef, CPid, _}} ->
-            #pool{free_pids = Free, in_use_count = NumInUse,
-                  free_count = NumFree} = Pool,
-            Pool1 = Pool#pool{in_use_count = NumInUse - 1},
-            Entry = {MRef, free, os:timestamp()},
-            Pool2 = Pool1#pool{all_members = store_all_members(Pid, Entry, AllMembers),
-                       consumer_to_pid = cpmap_remove(Pid, CPid,
-                                                      Pool1#pool.consumer_to_pid)},
-            case queue:out(QueuedRequestors) of
-                {empty, _ } ->
-                    Pool2#pool{free_pids = [Pid | Free], free_count = NumFree + 1};
-                {{value, {From = {APid, _}, TRef}}, NewQueuedRequestors} when is_pid(APid) ->
-                    reply_to_queued_requestor(TRef, Pid, From, NewQueuedRequestors, Pool2)
+            error_logger:info_msg("ConsumerPid ~p is no longer alive will check if Pid ~p is alive before returning",[CPid, Pid]),
+            %% check if the connection is alive before returning it.
+            case is_process_alive(Pid) of
+                true ->
+                    error_logger:info_msg("Pid ~p is alive, returning to pool ~p", [Pid, Pool]),
+                    #pool{free_pids = Free, in_use_count = NumInUse,
+                      free_count = NumFree} = Pool,
+                    Pool1 = Pool#pool{in_use_count = NumInUse - 1},
+                    Entry = {MRef, free, os:timestamp()},
+                    Pool2 = Pool1#pool{all_members = store_all_members(Pid, Entry, AllMembers),
+                               consumer_to_pid = cpmap_remove(Pid, CPid,
+                                                              Pool1#pool.consumer_to_pid)},
+                    error_logger:info_msg("Pid ~p added to pool ~p", [Pid, Pool2]),
+                    case queue:out(QueuedRequestors) of
+                        {empty, _ } ->
+                            Pool2#pool{free_pids = [Pid | Free], free_count = NumFree + 1};
+                        {{value, {From = {APid, _}, TRef}}, NewQueuedRequestors} when is_pid(APid) ->
+                            reply_to_queued_requestor(TRef, Pid, From, NewQueuedRequestors, Pool2)
+                    end;
+                false ->
+                    error_logger:info_msg("Pid ~p is dead will be removed from the pool ~p", [Pid, Pool]),
```

Comment:
For this log message, let's also include who it is/was being consumed by (the CPid)?